internal/services/v1/permissions.go

```diff
@@ -535,86 +536,157 @@ func (ps *permissionServer) LookupSubjects(req *v1.LookupSubjectsRequest, resp v
 	}
 	usagemetrics.SetInContext(ctx, respMetadata)
 
-	stream := dispatchpkg.NewHandlingDispatchStream(ctx, func(result *dispatch.DispatchLookupSubjectsResponse) error {
-		foundSubjects, ok := result.FoundSubjectsByResourceId[req.Resource.ObjectId]
-		if !ok {
-			return fmt.Errorf("missing resource ID in returned LS")
+	var currentCursor *dispatch.Cursor
+	remainingConcreteLimit := 0
+
+	lsRequestHash, err := computeLSRequestHash(req)
+	if err != nil {
+		return ps.rewriteError(ctx, err)
+	}
+
+	if req.OptionalCursor != nil {
+		decodedCursor, err := cursor.DecodeToDispatchCursor(req.OptionalCursor, lsRequestHash)
+		if err != nil {
+			return ps.rewriteError(ctx, err)
 		}
+		currentCursor = decodedCursor
+	}
+
+	if req.OptionalConcreteLimit > 0 {
+		remainingConcreteLimit = int(req.OptionalConcreteLimit)
+	}
+
+	internalResponseCursor := &impl.DecodedCursor{
+		VersionOneof: &impl.DecodedCursor_V1{
+			V1: &impl.V1Cursor{
+				Revision:              atRevision.String(),
+				CallAndParametersHash: lsRequestHash,
+			},
+		},
+	}
+
+	for {
+		countSubjectsFound := 0
 
-		for _, foundSubject := range foundSubjects.FoundSubjects {
-			excludedSubjectIDs := make([]string, 0, len(foundSubject.ExcludedSubjects))
-			for _, excludedSubject := range foundSubject.ExcludedSubjects {
-				excludedSubjectIDs = append(excludedSubjectIDs, excludedSubject.SubjectId)
+		ctxWithCancel, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		stream := dispatchpkg.NewHandlingDispatchStream(ctxWithCancel, func(result *dispatch.DispatchLookupSubjectsResponse) error {
+			foundSubjects, ok := result.FoundSubjectsByResourceId[req.Resource.ObjectId]
+			if !ok {
+				return fmt.Errorf("missing resource ID in returned LS")
 			}
 
-			excludedSubjects := make([]*v1.ResolvedSubject, 0, len(foundSubject.ExcludedSubjects))
-			for _, excludedSubject := range foundSubject.ExcludedSubjects {
-				resolvedExcludedSubject, err := foundSubjectToResolvedSubject(ctx, excludedSubject, caveatContext, ds)
+			for _, foundSubject := range foundSubjects.FoundSubjects {
+				// Skip wildcards if requested they be skipped.
+				if req.WildcardOption == v1.LookupSubjectsRequest_WILDCARD_OPTION_EXCLUDE_WILDCARDS && foundSubject.SubjectId == tuple.PublicWildcard {
+					continue
+				}
+
+				excludedSubjectIDs := make([]string, 0, len(foundSubject.ExcludedSubjects))
+				excludedSubjects := make([]*v1.ResolvedSubject, 0, len(foundSubject.ExcludedSubjects))
+				for _, excludedSubject := range foundSubject.ExcludedSubjects {
```

Comment:
We are missing tests where the dispatched operation returns excluded subjects. Ideally, we also test excluded subjects with caveats.

Reply:
Added; I came up with a string syntax to make writing the tests easier; let me know if you want any changes to that

Reply:
looks great!